### DISCUSSION TITLE
[vite-node] How to properly shutdown an HTTP server before script reload with `-w`?

### DISCUSSION THREAD

**Discussion by Azarattum** [vitest-dev/vitest#1738] [Q&A] [answered]
Hello! I have a question about `vite-node`.

I wonder how do I hot reload an http server? Right now it just reloads my script on save with `-w` mode. But the server from the previous execution of the script **keeps running**! So it results in port is used error. Is there a way to tell in a script when vite-node is about to reload, so I can gracefully stop the server? Or maybe there’s a way of reloading the server scripts without restarting the server itself?

Is there such a feature at all? Maybe I should open a feature request?

_btw, I'm using both http server and a web socket server in my app. I need them both to shutdown gracefully before reload. Also I don't need any client side HMR. My project is just an API._

**Comment by jarrodpayne** [comment]
@Azarattum did you ever figure out anything regarding this?

**Reply by Azarattum** [reply, marked as answer]
This is how I’ve implemented it in my project. It’s not super reliable, but works 80% of the time which is good enough for me.

```ts
if (import.meta.hot) await import.meta.hot.data.stopping;
if (import.meta.hot) {
  let reload = () => (info("Performing an HMR reload..."), stop());
  import.meta.hot.on("vite:beforeFullReload", () => {
    const stopping = reload();
    reload = () => Promise.resolve();
    if (import.meta.hot) import.meta.hot.data.stopping = stopping;
  });
}
```

Where `stop()` function stops and cleans up my HTTP server. I run this code when my application start.

For more context you can check out [the full code that I use in my project](https://github.com/Azarattum/Amadeus/blob/main/apps/core/status/manage.ts).

**Comment by josias-r** [comment]
Running into the same issue
```
[vite-node] Failed to execute file: 
 Error: listen EADDRINUSE: address already in use :::9999
```

Just gonna leave this here, so the error gets indexed by Google, so this discussion is easier to find.

I tried `vite-node --watch --no-hmr index.ts`, with **no luck**. So I'm hoping for more luck with the approach @Azarattum has taken


**Update:**
For my very similar use case, to @Azarattum, I am closing my WS and HTTP Servers like this:
```ts
import type { Server as HttpServer } from 'http';
import type { Server as IoServer } from 'socket.io';

function stopHttpAndSocketServers(ioServer: IoServer, httpServer: HttpServer) {
	ioServer.close();
	httpServer.close();
}

export async function registerViteHmrServerRestart(ioServer: IoServer, httpServer: HttpServer) {
	if (import.meta.hot) await import.meta.hot.data.stopping;
	if (import.meta.hot) {
		let reload = () => (
			console.info('Performing an HMR reload...'), stopHttpAndSocketServers(ioServer, httpServer)
		);
		import.meta.hot.on('vite:beforeFullReload', () => {
			const stopping = reload();
			reload = () => Promise.resolve();
			if (import.meta.hot) import.meta.hot.data.stopping = stopping;
		});
	}
}
```
my root `index.ts` file:
```ts
import { registerViteHmrServerRestart } from './vite-hmr-restart';

// ... setup servers

registerViteHmrServerRestart(io, server);
```

**Reply by simplecommerce** [reply]
This might not be a big issue or deal but have you guys noticed that if you save a file multiple times quickly it can cause an issue where multiple instances are trying to start up at the same time?

 I am trying to fix it on my end to prevent this but so far no luck even with the suggestions here.
